bin/generate

```diff
@@ -8,48 +8,56 @@ let generics = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M",
 
 extension Array {
   subscript(safe index: Int) -> Element? {
-    return indices ~= index ? self[index] : .None
+    return indices ~= index ? self[index] : .none
   }
 }
 
-func genericTypeForPosition(x: Int) -> String {
+func genericType(for position: Int) -> String {
   let max = generics.count
-  switch x {
-  case _ where x < max: return generics[x % max]
-  default: return generics[x / max - 1] + generics[x % max]
+  switch position {
+  case _ where position < max: return generics[position % max]
+  default: return generics[position / max - 1] + generics[position % max]
   }
 }
 
-func commaConcat(xs: [String]) -> String {
-  return xs.joinWithSeparator(", ")
+func commaConcat(_ xs: [String]) -> String {
+  return xs.joined(separator: ", ")
 }
 
-func curryDefinitionGenerator(arguments arguments: Int) -> String {
-  let genericParameters = (0..<arguments).map(genericTypeForPosition) // ["A", "B", "C"]
-  let genericTypeDefinition = "<\(commaConcat(genericParameters))>" // "<A, B, C>"
+func singleParameterFunctions(_ xs: [String]) -> String {
+  guard let first = xs.first else { fatalError("Attempted to nest functions with no arguments") }
+  guard xs.last != first else { return first }
+  let remainder = Array(xs.dropFirst())
+  return "(\(first)) -> \(singleParameterFunctions(remainder))"
+}
+
+func curryDefinitionGenerator(arguments: Int) -> String {
+  let genericParameters = (0..<arguments).map(genericType) // ["A", "B", "C", "D"]
+  let genericTypeDefinition = "<\(commaConcat(genericParameters))>" // "<A, B, C, D>"
 
-  let inputParameters = Array(genericParameters[0..<arguments - 1]) // ["A", "B"]
-  let lowerFunctionArguments = inputParameters.map { "`\($0.lowercaseString)`" } // ["`a`", "`b`"]
-  let returnType = genericParameters.last! // "C"
+  let inputParameters = Array(genericParameters[0..<arguments - 1]) // ["A", "B", "C"]
+  let lowerFunctionArguments = inputParameters.map { "`\($0.lowercased())`" } // ["`a`", "`b`", "`c`"]
+  let returnType = genericParameters.last! // "D"
 
-  let functionArguments = "(\(commaConcat(inputParameters)))" // "(A, B)"
-  let returnFunctionArguments = inputParameters.joinWithSeparator(" -> ") // " A -> B"
-  let innerFunctionArguments = commaConcat(lowerFunctionArguments) // "`a`, `b`"
+  let functionArguments = "(\(commaConcat(inputParameters)))" // "(A, B, C)"
+  let returnFunction = singleParameterFunctions(genericParameters) // " (A) -> (B) -> (C) -> D"
+  let innerFunctionArguments = commaConcat(lowerFunctionArguments) // "`a`, `b`, `c`"
 
-  let functionDefinition = "function(\(innerFunctionArguments))" // return function(`a`, `b`)
+  let functionDefinition = "function(\(innerFunctionArguments))" // return function(`a`, `b`, `c`)
```

Comment:
Can we update this code comment to remove `"return"`? Doesn't look like it's actually added here.